### PR TITLE
Codeocean import export

### PR DESCRIPTION
**Pull request by kkoehn**
Implements interfaces and endpoints to communicate with Codeocean

This PR should be merged in main PR before master

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -55,16 +54,18 @@ gem 'cancancan'
...
 gem 'better_errors' # !!!Important!!! move this to the development group before codeharbor goes really productive!!!Important!!!
+gem 'config'
+gem 'faraday'
 gem 'font-awesome-rails', '>= 4.7.0.5'
 gem 'jquery-ui-rails', '>= 6.0.1'
 gem 'less-rails', '>= 3.0.0' # Sprockets (what Rails 3.1 uses for its asset pipeline) supports LESS
 gem 'nested_form_fields', '>= 0.8.2'
 gem 'nokogiri'
-gem 'oauth2', '~> 1.4.0'
+gem 'proforma', git: 'https://github.com/openHPI/proforma.git', tag: 'v0.3.2'
```
Comment (MrSerth) [inline Gemfile]: Same as #122: Should we consider publishing the gem and use that version?

File: app/assets/javascripts/exercises.coffee [hunk 2]
```diff
@@ -199,10 +199,16 @@ ready =->
...
-    return
 
-  $('#file_upload').on 'change', ->
-    fullPath = document.getElementById('file_upload').value
+  $('#import-exercise-button').on 'click', (e) ->
+    e.preventDefault()
+    e.stopPropagation()
+    console.log('click')
+    importExerciseStart()
+    # return false
```
Comment (MrSerth) [inline app/assets/javascripts/exercises.coffee]: No longer required, right?
```suggestion
```

File: app/assets/javascripts/exercises.coffee [hunk 3]
```diff
@@ -199,10 +199,16 @@ ready =->
...
   $('#xml-import').on 'click', (e) ->
     e.stopPropagation()
-    return
 
-  $('#file_upload').on 'change', ->
-    fullPath = document.getElementById('file_upload').value
+  $('#import-exercise-button').on 'click', (e) ->
+    e.preventDefault()
+    e.stopPropagation()
+    console.log('click')
```
Comment (MrSerth) [inline app/assets/javascripts/exercises.coffee]: 🚧 
```suggestion
```

File: app/assets/javascripts/exercises.coffee [hunk 4]
```diff
@@ -302,7 +308,7 @@ ready =->
       return
     related_exercises = $(this).siblings('.content').children()
     last_shown_element = $(related_exercises).filter('.displayed').last()
-    console.log(last_shown_element)
+    # console.log(last_shown_element)
```
Comment (MrSerth) [inline app/assets/javascripts/exercises.coffee]:
```suggestion
```

File: app/assets/javascripts/exercises.coffee [hunk 5]
```diff
@@ -418,7 +424,142 @@ ready =->
...
     $('.primary-checkbox')[0].click()
 
 $(document).on('turbolinks:load', ready)
+
+checkExecutionEnvironment = () ->
+  executionEnvironment = $('#exercise_execution_environment_id').val()
+  visibilityPrivate = $('input:radio:checked[name="exercise[private]"]').val()
+
+  if executionEnvironment == '' && visibilityPrivate == 'false'
+    console.log 'show'
```
Comment (MrSerth) [inline app/assets/javascripts/exercises.coffee]:
```suggestion
```

File: app/assets/javascripts/exercises.coffee [hunk 6]
```diff
@@ -418,7 +424,142 @@ ready =->
...
+    $('.execution-environment-error').removeClass('hidden')
+    $('.execution-environment-error').addClass('errors')
+    $('.execution-environment-error').parent().find('.form-label').addClass('errors')
+  else
+    console.log 'hide'
```
Comment (MrSerth) [inline app/assets/javascripts/exercises.coffee]:
```suggestion
```

File: app/assets/javascripts/exercises.coffee [hunk 7]
```diff
@@ -418,7 +424,142 @@ ready =->
...
+    $('.execution-environment-error').addClass('hidden')
+    $('.execution-environment-error').removeClass('errors')
+    $('.execution-environment-error').parent().find('.form-label').removeClass('errors')
+
+importExerciseStart = () -> #maybe unobstrusive like export?
+  # console.log('start')
```
Comment (MrSerth) [inline app/assets/javascripts/exercises.coffee]:
```suggestion
```

File: app/assets/javascripts/exercises.coffee [hunk 8]
```diff
@@ -418,7 +424,142 @@ ready =->
...
+  formData.append('zip_file', document.getElementById('exercise_import').files[0])
+
+  $.ajax({
+    type: 'POST',
+    url: '/exercises/import_exercise_start',
+    data: formData,
+    processData: false,
+    contentType: false,
+    # success: (response) ->
+      # console.log(response)
```
Comment (MrSerth) [inline app/assets/javascripts/exercises.coffee]: Is here something missing? This request is used to stat the import process while the other one below is used to grab the import status, right?

File: app/assets/javascripts/exercises.coffee [hunk 9]
```diff
@@ -418,7 +424,142 @@ ready =->
...
+  $messageDiv.html('Checking if the exercise exists on Codeocean.')
+  $actionsDiv.html('<div class="spinner-border"></div>')
+
+  $.ajax({
+    type: 'POST',
+    url: '/exercises/' + exerciseID + '/export_external_check',
+    data: {account_link: accountLinkID},
+    dataType: 'json',
+    success: (response) ->
+      # if response.error
```
Comment (MrSerth) [inline app/assets/javascripts/exercises.coffee]: same as above

File: app/assets/javascripts/exercises.coffee [hunk 10]
```diff
@@ -418,7 +424,142 @@ ready =->
...
+    success: (response) ->
+      $messageDiv.html response.message
+      $actionsDiv.html response.actions
+
+      if response.status == 'success'
+        $exerciseDiv.addClass 'import-export-success'
+      else
+        $exerciseDiv.addClass 'import-export-failure'
+
+      # checkExportDialog()
```
Comment (MrSerth) [inline app/assets/javascripts/exercises.coffee]: method no longer used? Here and a few lines below.

File: app/assets/stylesheets/exercises.scss [hunk 11]
```diff
@@ -673,3 +674,69 @@ input[type="radio"]:checked+label {
...
+  color: #080;
+}
+
+.confirm-icon-alt {
+  color: #008;
+}
+
+.abort-icon {
+  color: #800;
+  // width: 12px;
```
Comment (MrSerth) [inline app/assets/stylesheets/exercises.scss]:
```suggestion
```

File: app/assets/javascripts/exercises.coffee (hunk 8 above)
Comment (kkoehn) [reply app/assets/javascripts/exercises.coffee]: This one starts the dialog while `import_exercise_confirm` starts the actual import process.

Comment (MrSerth) [reply app/assets/javascripts/exercises.coffee]: Alright, I see. Thanks for clarification :)

File: config/locales/en.yml [hunk 12]
```diff
@@ -19,7 +19,13 @@
 # To learn more, please read the Rails Internationalization guide
 # available at http://guides.rubyonrails.org/i18n.html.
 
+
+# activerecord.errors.models.exercise.attributes.execution_environment.blank
```
Comment (MrSerth) [inline config/locales/en.yml]: Do we still need that?
```suggestion
```

**Pull request merged**